packages/backend/src/invoice/routes.ts

```diff
@@ -0,0 +1,147 @@
+import { StreamServer } from '@interledger/stream-receiver'
+import { Logger } from 'pino'
+import { validateId } from '../shared/utils'
+import { AppContext } from '../app'
+import { IAppConfig } from '../config/app'
+import { AccountService } from '../account/service'
+import { InvoiceService } from './service'
+import { Invoice } from './model'
+
+// Don't allow creating an invoice too far out. Invoices with no payments before they expire are cleaned up, since invoice creation is unauthenticated.
+// TODO what is a good default value for this?
+export const MAX_EXPIRY = 24 * 60 * 60 * 1000 // milliseconds
+
+interface ServiceDependencies {
+  config: IAppConfig
+  logger: Logger
+  accountService: AccountService
+  invoiceService: InvoiceService
+  streamServer: StreamServer
+}
+
+export interface InvoiceRoutes {
+  get(ctx: AppContext): Promise<void>
+  create(ctx: AppContext): Promise<void>
+}
+
+export function createInvoiceRoutes(deps_: ServiceDependencies): InvoiceRoutes {
+  const logger = deps_.logger.child({
+    service: 'InvoiceRoutes'
+  })
+  const deps = { ...deps_, logger }
+  return {
+    get: (ctx: AppContext) => getInvoice(deps, ctx),
+    create: (ctx: AppContext) => createInvoice(deps, ctx)
+  }
+}
+
+// Spec: https://docs.openpayments.dev/invoices#get
+async function getInvoice(
+  deps: ServiceDependencies,
+  ctx: AppContext
+): Promise<void> {
+  const { invoiceId } = ctx.params
+  ctx.assert(validateId(invoiceId), 400, 'invalid id')
+  const acceptJSON = ctx.accepts('application/json')
+  const acceptStream = ctx.accepts('application/ilp-stream+json')
+  ctx.assert(acceptJSON || acceptStream, 406)
+
+  const invoice = await deps.invoiceService.get(invoiceId)
+  if (!invoice) return ctx.throw(404)
+
+  const amountReceived = await deps.accountService.getBalance(invoice.accountId)
+  if (amountReceived === undefined) {
+    deps.logger.error(
+      { account: invoice.accountId, invoice: invoice.id },
+      'balance not found'
+    )
+    return ctx.throw(500)
+  }
+
+  const body = invoiceToBody(deps, invoice, amountReceived)
+  ctx.body = body
+  if (!acceptStream) return
+
+  const { ilpAddress, sharedSecret } = deps.streamServer.generateCredentials({
+    paymentTag: invoice.accountId,
+    // TODO receipt support on invoices?
+    //receiptSetup:
+    //  nonce && secret
+    //    ? {
+    //        nonce: Buffer.from(nonce.toString(), 'base64'),
+    //        secret: Buffer.from(secret.toString(), 'base64')
+    //      }
+    //    : undefined,
+    asset: {
+      code: invoice.account.asset.code,
+      scale: invoice.account.asset.scale
+    }
+  })
+
+  body['ilpAddress'] = ilpAddress
+  body['sharedSecret'] = sharedSecret
```

Comment:
Does this need to be base64 encoded like in spsp?
https://github.com/interledger/rafiki/blob/main/packages/backend/src/spsp/service.ts#L98